doc/source/tutorials/sensor-graph-advanced.rst

```diff
@@ -0,0 +1,251 @@
+
+	This will fire exactly once for every 60 readings added to `counter 1`
+	because each time it runs it will reset the count on its input StreamWalker
+	back to zero.
+
+Different Kinds of Streams 
+##########################
+
+There are currently X different classes of streams.  Their own differences are 
+in how many past readings values are remembered and whether a count is kept
+of how many readings have been pushed to the stream. 
+
+Buffered Streams
+	Buffered streams can be considered as normal FIFOs.  All readings pushed to
+	a buffered stream are remembered until the device runs out of storage space
+	and the count of available readings corresponds with the number of readings
+	that have been pushed to the stream with each pop() operation returning the
+	next oldest reading.
+
+Unbuffered Streams
+	Unbuffered streams only every store 1 value.  They have no space to store
+	historical data and they also don't lie to you about how many readings are
+	available so an unbuffered stream can only ever have a count of 0 or 1
+	depending on whether it has data available or not.
+
+Counter Streams
+	Counter streams are unbuffered so they only store a single reading, however,
+	they keep an accurate count of how many times they have been pushed to and
+	allow you to pop from them that many times, each time returning the same 
+	latest value that was last pushed.  Counter streams are primarily useful
+	for creating efficient timers but their values are typically not used, just
+	their counts. 
+
+Input Streams
+	Input streams are the global inputs to a sensor graph.  They are the roots
+	of the processing graph.  The only entry points for new data into a sensor
+	graph are inputs.  They are unbuffered.
+
+Output Streams
+	Output streams are buffered streams but stored in a different region of 
+	persistent storage from buffered streams so that overflowing the buffered
+	storage region does not overflow the output storage.  As the name suggests,
+	Output streams typically represent the outputs of a device that should be 
+	saved historically.
+
+Constant Streams
+	Constant streams always return a constant value.  They can never be 
+	exhausted and are useful for two primary purposes.  The first is to embed
+	constant data in a sensor graph like what RPCs to call.  The second is to
+	create latches that are used to derive timers gated on specific events.
+
+	For example, if the user creates a `when connected` block that should call
+	an RPC every second while a user is connected to the device, internally a
+	constant stream is used to create a latch that is 1 when the user is
+	connected and 0 otherwise.  This is combined with a 1 second clock to 
+	create a derived 1 second clock that is only active when a user is
+	connected.
+
+Users need to explicitly specify the types of each stream they want to allocate
+since it's not possible for the SensorGraph compiler to infer which would be 
+most appropriate in most cases.
```

Comment:
consider changing 'most appropriate' to 'appropriate' to avoid using 'most' repeatedly